### PR TITLE
fixes #1152 and an odd corner case

### LINKED ISSUE #1152: Re-ordering downloads in progress tab no longer works
Set a low download rate if you have a fast connection and set the limit to one download. Add three downloads, drag the bottom (# 3) download to the top, pause the current download (# 1). The top entry (# 3) should begin downloading but doesn't until entry # 2 finishes. The original order is retained even when moving them.

Introduced in #1054.

### PR DESCRIPTION
**Pull request by blushingpenguin**
To fix the task queue re-ordering, this basically reverts the chunk of #1054 that added a second thread safe tasks queue and goes back to using the Gtk.ListStore to contain the tasks work queue. To workaround the problem of the background download threads triggering exceptions when concurrently accessing the list store while the main ui thread is manipulating it, only the main ui thread accesses the list store, and background worker threads queue a request for a task to the main thread by adding an idle function which dequeues a task, then sends the result back to the background thread (via the DequeRequest class).

This also fixes the corner case of downloading a task that no longer exists. That can be triggered by downloading some episodes, then in the tasks list cancelling one of the downloads. Once cancellation has finished, cancel all other downloads, then quickly right click on the cancelled task to get the context menu. Wait until the other tasks have been cancelled and the tasks list has been cleared, then choose to download the cancelled task -- at this point the task is in the queued state, not in the list (and hence not in the queue) but attached to the episode,so it becomes impossible to download the episode that the task is attached to. This is fixed by detecting the case of the task not being in the tasks list and adding it back in if necessary.

### CODE REVIEW

**Comment by auouymous** [comment]
Is the second issue https://github.com/gpodder/gpodder/pull/1149#issuecomment-909259747?

**Comment by blushingpenguin** [comment]
@auouymous Yes, that is the second issue

File: src/gpodder/gtkui/main.py [hunk 1]
```diff
@@ -1177,7 +1177,7 @@ def update_downloads_list(self, can_call_cleanup=True):
 
                 task = row[self.download_status_model.C_TASK]
                 speed, size, status, progress, activity = task.speed, task.total_size, task.status, task.progress, task.activity
-                logger.info("%s: %f", task.episode.title, progress)
+                # logger.info("%s: %f", task.episode.title, progress)
```
Comment (auouymous) [inline src/gpodder/gtkui/main.py:1180]: This should be removed.

File: src/gpodder/gtkui/main.py [hunk 2]
```diff
@@ -2244,6 +2238,7 @@ def play_or_download(self, current_page=None):
                     else:
                         streaming_possible |= self.streaming_possible(episode)
                         can_download = True
+                logger.info('%s - can_download=%s, can_play=%s, can_delete = %s' % (episode.description, can_download, can_play, can_delete))
```
Comment (auouymous) [inline src/gpodder/gtkui/main.py:2253]: You removed the other debug line in a later patch but missed this one.

File: src/gpodder/gtkui/main.py [hunk 3]
```diff
@@ -2210,7 +2207,6 @@ def play_or_download(self, current_page=None):
             return (False, False, False, False, False)
 
         (can_play, can_download, can_cancel, can_delete) = (False,) * 4
-        (is_played, is_locked) = (False,) * 2
```
Comment (auouymous) [inline src/gpodder/gtkui/main.py:2213]: Removing these unused variables is unrelated to the fix and should happen in a separate PR.

**Review by auouymous — CHANGES_REQUESTED** [review]

File: src/gpodder/download.py [hunk 4]
```diff
@@ -936,14 +936,6 @@ def run(self):
             self.error_message = _('Error: %s') % (str(e),)
 
         with self:
-            if result == DownloadTask.FAILED:
```
Comment (auouymous) [inline src/gpodder/download.py:919]: Is the moving of this block related to the fix or simply to make the normal path faster than the failure path?

**Review by auouymous — CHANGES_REQUESTED** [review]

**Comment by auouymous** [comment]
I'm unable to cancel failed downloads (the 'ccc' episodes at http://borg.qzx.com/test-rss.php always fail).

File: src/gpodder/download.py (hunk 4 above)
Comment (blushingpenguin) [reply src/gpodder/download.py:919]: it was meant to make things easier to read by making the paths less complicated, I've further refactored the code so it is now:

if SUCCESS:
  do things
  return True

clean up after pause/cancel/fail

if FAIL:
  things
elif status = PAUSING:
  self.status = PAUSED
elif status = CANCELLING:
  self.status = CANCELLED

I think that's even clearer as there are less paths through the code.  (Happy to change this back if you thought it was clearer before)

File: src/gpodder/gtkui/main.py (hunk 3 above)
Comment (blushingpenguin) [reply src/gpodder/gtkui/main.py]: I've split this off into https://github.com/gpodder/gpodder/pull/1158

**Comment by blushingpenguin** [comment]
> I'm unable to cancel failed downloads (the 'ccc' episodes at http://borg.qzx.com/test-rss.php always fail).
fixed

**Comment by blushingpenguin** [comment]
Thanks for looking at this

**Comment by auouymous** [comment]
Thanks for fixing these issues.

**Comment by elelay** [comment]
Thank you both for your hard work making things work again after the switch to gio!
